File: web/src/core/clock.ts
/**
 * ChartClock — the one honest clock for a play session.
 *
 * The pygame build anchors on `time.perf_counter()` and slews toward
 * `mixer.music.get_pos()`, because the OS audio stack can stall and the mixer's
 * idea of "now" is the only truth available.  The browser hands us that truth
 * directly: `AudioContext.currentTime` advances with the audio hardware's sample
 * clock, so chart time is read off the audio graph and there is nothing to
 * correct.  `check_drift` from clock.py has no counterpart here by design.
 *
 * Before the music starts (the count-in), the clock runs off `performance.now()`
 * anchored to the same instant, so the lead-in bars fall on a clock that is
 * continuous with the one the song will use.
 *
 * Everything that judges or draws a note asks this clock.  Pause lives inside it,
 * so a paused run never drifts.
 */

export class ChartClock {
  private ctx: AudioContext
  /** ctx.currentTime at chart time 0 */
  private anchor: number | null = null
  private pausedTotal = 0
  private pauseStart: number | null = null
  private offset: number

  constructor(ctx: AudioContext, deviceOffsetMs = 0) {
    this.ctx = ctx
    this.offset = deviceOffsetMs / 1000
  }

  // ── lifecycle ────────────────────────────────────────────────────────────
  start(): void {
    this.anchor = this.ctx.currentTime
    this.pausedTotal = 0
    this.pauseStart = null
  }

  get started(): boolean { return this.anchor !== null }
  get paused(): boolean { return this.pauseStart !== null }

  pause(): void {
    if (this.anchor === null || this.pauseStart !== null) return
    this.pauseStart = this.ctx.currentTime
  }

  resume(): void {
    if (this.pauseStart === null) return
    this.pausedTotal += this.ctx.currentTime - this.pauseStart
    this.pauseStart = null
  }

  // ── reading ──────────────────────────────────────────────────────────────
  /** Chart time without the device offset. */
  raw(): number {
    if (this.anchor === null) return 0
    const now = this.pauseStart !== null ? this.pauseStart : this.ctx.currentTime
    return now - this.anchor - this.pausedTotal
  }

  /** Chart time in seconds: what judgments and rendering use. */
  now(): number {
    return this.raw() + this.offset
  }

  /**
   * Chart time for a press that happened at `eventTimeMs` (a DOM
   * `KeyboardEvent.timeStamp`, i.e. milliseconds on the `performance.now()`
   * timeline).
   *
   * This is the whole reason the web build judges better than the pygame one:
   * a key pressed mid-frame is judged at the instant it was pressed, not at the
   * instant the frame got around to reading it.  A 16 ms frame used to cost up
   * to 16 ms of judgment error; here it costs none.
   */
  atEvent(eventTimeMs: number): number {
    const lagS = (performance.now() - eventTimeMs) / 1000
    return this.now() - Math.max(0, lagS)
  }

  setDeviceOffset(ms: number): void { this.offset = ms / 1000 }

  /**
   * Re-anchor so that `raw()` equals `chartT` right now.
   *
   * Called the instant the music actually starts (after the count-in), so the
   * lead-in bars can never leave the chart a frame out from the song.  `at` is
   * the AudioContext time the source was scheduled to begin, which is known
   * exactly because we scheduled it.
   */
  rebase(chartT: number, at?: number): void {
    if (this.anchor === null) return
    const when = at ?? this.ctx.currentTime
    this.anchor = when - this.pausedTotal - chartT
  }
}

File: game/clock.py
"""
ChartClock — the one honest clock for a play session.

Anchored the instant the song starts (call ``start()`` right after
``pygame.mixer.music.play()``), it returns *chart time*: seconds since the
song began, excluding pauses, plus the device offset from settings.

Everything that judges or draws a note asks this clock; nothing else reads
``time.perf_counter()`` directly.  Pause lives inside the clock so a paused
run never drifts.  An optional slew keeps chart time within a few ms of the
mixer's own position when the OS audio stack stalls.
"""
from __future__ import annotations

import time


class ChartClock:
    def __init__(self, device_offset_ms: float = 0.0, slew: bool = True) -> None:
        self._anchor: float | None = None
        self._paused_total: float = 0.0
        self._pause_start: float | None = None
        self._offset = device_offset_ms / 1000.0
        self._slew_enabled = slew
        self._slew: float = 0.0          # correction applied on top of raw time
        self._last_slew_check: float = 0.0

    # ── lifecycle ──────────────────────────────────────────────────────────
    def start(self) -> None:
        self._anchor = time.perf_counter()
        self._paused_total = 0.0
        self._pause_start = None
        self._slew = 0.0

    @property
    def started(self) -> bool:
        return self._anchor is not None

    @property
    def paused(self) -> bool:
        return self._pause_start is not None

    def pause(self) -> None:
        if self._anchor is None or self._pause_start is not None:
            return
        self._pause_start = time.perf_counter()

    def resume(self) -> None:
        if self._pause_start is None:
            return
        self._paused_total += time.perf_counter() - self._pause_start
        self._pause_start = None

    # ── reading ───────────────────────────────────────────────────────────
    def raw(self) -> float:
        """Chart time without the device offset or slew."""
        if self._anchor is None:
            return 0.0
        now = self._pause_start if self._pause_start is not None else time.perf_counter()
        return now - self._anchor - self._paused_total

    def now(self) -> float:
        """Chart time in seconds (what judgments and rendering use)."""
        return self.raw() + self._offset + self._slew

    def set_device_offset(self, ms: float) -> None:
        self._offset = ms / 1000.0

    def rebase(self, chart_t: float) -> None:
        """Re-anchor so that ``raw()`` equals ``chart_t`` right now.

        Called the instant the music actually starts (after the count-in), so
        the lead-in bars can never leave the chart a frame out from the song.
        """
        if self._anchor is None:
            return
        self._anchor = time.perf_counter() - self._paused_total - chart_t
        self._slew = 0.0

    # ── drift correction ──────────────────────────────────────────────────
    def check_drift(self, mixer_pos_s: float | None, max_step: float = 0.002) -> None:
        """Nudge chart time toward the mixer position, at most ``max_step`` per call.

        ``mixer_pos_s`` is ``pygame.mixer.music.get_pos() / 1000`` (seconds the music
        has actually played).  Only corrects when the drift exceeds 15 ms, so the
        normal jitter of get_pos() never moves the clock.
        """
        if not self._slew_enabled or mixer_pos_s is None or mixer_pos_s < 0:
            return
        if self._anchor is None or self._pause_start is not None:
            return
        drift = mixer_pos_s - self.raw() - self._slew
        if abs(drift) > 0.015:
            step = max(-max_step, min(max_step, drift))
            self._slew += step
